authentication-client/authentication-client-java/src/main/java/co/cask/cdap/security/authentication/client/basic/BasicAuthenticationClient.java

```
@@ -0,0 +1,183 @@
+    auth = auth.replaceAll("(\r|\n)", StringUtils.EMPTY);
+    getRequest.addHeader(HttpHeaders.AUTHORIZATION, AUTHENTICATION_HEADER_PREFIX_BASIC + auth);
+
+    HttpResponse httpResponse = httpClient.execute(getRequest);
+    RestClientUtils.verifyResponseCode(httpResponse);
+
+    JsonElement jsonElement = RestClientUtils.toJsonElement(httpResponse.getEntity());
+    Map<String, String> responseMap = GSON.fromJson(jsonElement, new TypeToken<Map<String, String>>() {
+    }.getType());
+    String tokenValue = responseMap.get(ACCESS_TOKEN_KEY);
+    String tokenType = responseMap.get(TOKEN_TYPE_KEY);
+    if (StringUtils.isEmpty(tokenValue) || StringUtils.isEmpty(tokenType)) {
+      throw new IOException("Unexpected response was received from the authentication server.");
+    }
+
+    String expiresInStr = responseMap.get(EXPIRES_IN_KEY);
+    Long expiresIn = StringUtils.isNotEmpty(expiresInStr) ? Long.valueOf(expiresInStr) : Long.valueOf(0);
+    return new AccessToken(tokenValue, expiresIn, tokenType);
+  }
+
+  private String getAuthURL() throws IOException {
+    if (baseUrl == null) {
+      throw new IllegalStateException("Base authentication client is not configured!");
+    }
+
+    LOG.debug("Try to get the authentication URI from the gateway server: {}.", baseUrl);
+    String result = StringUtils.EMPTY;
+    HttpGet get = new HttpGet(baseUrl);
+    HttpResponse response = httpClient.execute(get);
+    if (response.getStatusLine().getStatusCode() == HttpStatus.SC_UNAUTHORIZED) {
+      JsonElement jsonElement = RestClientUtils.toJsonElement(response.getEntity());
+      Map<String, List<String>> responseMap = GSON.fromJson(jsonElement,
+                                                            new TypeToken<Map<String, List<String>>>() {
+                                                            }.getType());
+      List<String> uriList = responseMap.get(AUTH_URI_KEY);
+      if (uriList != null && !uriList.isEmpty()) {
```

If uriList is null or empty then we should throw an exception since we cannot find the authentication servers.